plugins/cryptolib/cryptolib.cpp

```diff
@@ -528,6 +525,149 @@ CRYPTOLIB_API bool CRYPTOLIB_CALL clPKIVerifySignatureBuff(const char * pkalgori
+        KeyCache::iterator it = keyCache.find(searchKey.str());
+        CLoadedKey * ret = nullptr;
+        if (it != keyCache.end())//exists in cache?
+        {
+            ret = (*it).second;
+        }
+        else
+        {
+            if (!isEmptyString(keyFS))
+            {
+                //Create CLoadedKey from filespec
+                if (isPublic)
+                    ret = loadPublicKeyFromFile(keyFS, passphrase);
+                else
+                    ret = loadPrivateKeyFromFile(keyFS, passphrase);
+            }
+            else
+            {
+                //Create CLoadedKey from key contents
+                if (isPublic)
+                    ret = loadPublicKeyFromMemory(keyBuff, passphrase);
+                else
+                    ret = loadPrivateKeyFromMemory(keyBuff, passphrase);
+            }
+
+            keyCache.emplace(searchKey.str(), ret);
+        }
+        return LINK(ret);
+    }
+};
+
+#else
+
+class CKeyCache : public CInterface
+{
+private:
+    bool        m_isPublic;
+    StringAttr  m_keyFS;
+    StringAttr  m_keyBuff;
+    StringAttr  m_passphrase;
+    CLoadedKey * m_loadedKey = nullptr;
+
+    inline bool sameString(const char * left, const char * right)
+    {
+        if (nullptr == left && nullptr == right)
+            return true;
+        if (nullptr == left || nullptr == right)
+            return false;
+        return strcmp(left, right) == 0;
+    }
+
+public:
+    virtual ~CKeyCache()
+    {
+        if (m_loadedKey)
+            m_loadedKey->Release();
+    }
+
+    CLoadedKey * getInstance(bool isPublic, const char * keyFS, const char * keyBuff, const char * passphrase)
+    {
+        if (isEmptyString(keyFS) && isEmptyString(keyBuff))
+            throw MakeStringException(-1, "Must specify a key filename or provide a key buffer");
+
+        if (isPublic != m_isPublic ||
+            !sameString(passphrase, m_passphrase.str())  ||
+            !sameString(keyFS, m_keyFS.str()) ||
+            !sameString(keyBuff, m_keyBuff.str()))
+        {
+            if (m_loadedKey)
+            {
+                m_loadedKey->Release();//release previous cached key
+                m_keyFS.clear();
+                m_keyBuff.clear();
+                m_passphrase.clear();
+            }
+            if (!isEmptyString(keyFS))
+            {
+                //Create CLoadedKey from filespec
+                if (isPublic)
+                    m_loadedKey = loadPublicKeyFromFile(keyFS, passphrase);
+                else
+                    m_loadedKey = loadPrivateKeyFromFile(keyFS, passphrase);
+            }
+            else
+            {
+                //Create CLoadedKey from key contents
+                if (isPublic)
+                    m_loadedKey = loadPublicKeyFromMemory(keyBuff, passphrase);
+                else
+                    m_loadedKey = loadPrivateKeyFromMemory(keyBuff, passphrase);
+            }
+            if (m_loadedKey)
```

Comment:
minor: Can this ever be null?  I don't think so, so it would be better not to check.

Reply:
The first time through it is null

Reply:
I don't think it is - it is initialized by the load functions above, and should never be null.